packages/abstractions/test/common/authentication/apiKeyAutenticationProvider.ts

```diff
@@ -0,0 +1,73 @@
+/**
+ * -------------------------------------------------------------------------------------------
+ * Copyright (c) Microsoft Corporation.  All Rights Reserved.  Licensed under the MIT License.
+ * See License in the project root for license information.
+ * -------------------------------------------------------------------------------------------
+ */
+
+import * as chai from "chai";
+
+import {
+  ApiKeyAuthenticationProvider,
+  ApiKeyLocation,
+} from "../../../src/authentication";
+import { RequestInformation } from "../../../src/requestInformation";
+const assert = chai.assert;
+
+describe("ApiKeyAuthenticationProvider", () => {
+  it("Throws on invalid initialization", () => {
+    assert.throws(
+      () =>
+        new ApiKeyAuthenticationProvider(
+          "",
+          "param",
+          ApiKeyLocation.QueryParameter
+        )
+    );
+    assert.throws(
+      () =>
+        new ApiKeyAuthenticationProvider(
+          "param",
+          "",
+          ApiKeyLocation.QueryParameter
+        )
+    );
+  });
+  it("Adds in query parameters", async () => {
+    const provider = new ApiKeyAuthenticationProvider(
+      "key",
+      "param",
+      ApiKeyLocation.QueryParameter
+    );
+    const request = new RequestInformation();
+    request.urlTemplate = "https://localhost{?param1}";
+    await provider.authenticateRequest(request);
+    assert.equal(request.URL, "https://localhost?param=key");
+    assert.isUndefined(request.headers["param"]);
+  });
+  it("Adds in query parameters with other parameters", async () => {
+    const provider = new ApiKeyAuthenticationProvider(
+      "key",
+      "param",
+      ApiKeyLocation.QueryParameter
+    );
+    const request = new RequestInformation();
+    request.urlTemplate = "https://localhost{?param1}";
+    request.queryParameters["param1"] = "value1";
+    await provider.authenticateRequest(request);
+    assert.equal(request.URL, "https://localhost?param1=value1&param=key");
+    assert.isUndefined(request.headers["param"]);
+  });
+  it("Adds in headers", async () => {
+    const provider = new ApiKeyAuthenticationProvider(
+      "key",
+      "param",
+      ApiKeyLocation.Header
+    );
+    const request = new RequestInformation();
+    request.urlTemplate = "https://localhost{?param1}";
+    await provider.authenticateRequest(request);
+    assert.equal(request.URL, "https://localhost");
+    assert.equal(request.headers["param"], "key");
+  });
```

Comment:
Can we add a test that validates the error throwing behavior of the authenticate request?

Reply:
done, also moved the defensive programing earlier in the flow.